### ISSUE TITLE
[DB IO managers] support when a new column is added to an asset

### ISSUE THREAD

**Issue by jamiedemaria** [dagster-io/dagster#13098]
Right now if you have a partitioned asset and you change the dataframe returned by the asset to have a new column, the IO manager will fail (specifically the snowflake pandas io manager will fail, still need to test the other io managers). This is because, for partitioned assets, we don't fully delete and replace the table for the asset materialization. Instead we drop the table (but retain the column schema) or delete the rows corresponding to the partition and then append the new dataframe. Since the new dataframe has a column that doesn't exist in the snowflake table, we see an error

Ideas of implementation
Two options (potentially more as well)

a) introspect on the dataframe, find the new column, map it to the DB column type, and manually create the column. We'd need to find a way to keep our mapping of python type to database type in sync.

b) copy the full table into memory, append the new partition data with the new column, drop the table, the load the new table into the DB. This would be super memory intensive for large tables, which is likely a deal breaker

Additional information
OG slack thread https://dagster.slack.com/archives/C01U954MEER/p1679514144698129

Message from the maintainers
Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by sryza** [comment]
I agree that (b) is a no-go, for the reason you listed.

**Comment by sryza** [comment]
(a) seems promising

**Comment by jamiedemaria** [comment]
This issue also exists with non-partitioned assets. Replication case

```
@asset
def a_plain_table() -> pd.DataFrame:
    return pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6],})
```
and materialize. Then update asset def to 
```
@asset
def a_plain_table() -> pd.DataFrame:
    return pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6], "c": [7, 8, 9]})
```
and materialization will fail

**Comment by DanMeyer** [comment]
I've been doing (b) manually. I have smaller tables than most (the largest one has 2-3 million records), so this hasn't been a problem performance-wise. However, I have a few Snowflake roles (some for user-facing apps) that only have access to a few specific tables (so I can't just grant select on all future tables in a schema), and I lose those permission grants when I drop the tables.

My best idea so far is include a statement granting those permissions as part of any such assets.

**Comment by malow106** [comment]
Hi, any plan for this issue ? Thx !

**Comment by geoHeil** [comment]
What about a not fully automatic mode but a configuration-driven approach where one can decide what should happen on a per-asset (metadata) level?

**Comment by milesgranger** [comment]
> What about a not fully automatic mode but a configuration-driven approach...

A (very) common situation in BI/DWH groups are sources changing schemas which are either probably good to know about or they break things, and that's something only these groups know how to properly handle, not dagster.

IMO, having a dependency injection to the DB IO Managers which can hand off the delta change in schema, and get back an acknowledgement on the action the manager should take (recreate, alter, error, ...) would be _very_ helpful.

Data point:  we have sources that frequently augment their existing data but rarely remove things in a breaking way. As it is now, this breaks things due to this missing feature; a callback to post a notification to Teams channel then tell the DB manager to go ahead with altering/recreating the table or alternatively alert on breaking changes and preventing the update would be glorious! 

Don't try to be too smart w/ this saving/comparing of schemas, automagic, without exposing this to the users.

**Comment by geoHeil** [comment]
Agreed. https://docs.confluent.io/platform/current/schema-registry/fundamentals/schema-evolution.html is a good read on the complexity

**Comment by Superjisan** [comment]
This would be a very useful feature for us!
